File: 2022/Python/Day-3/Part-2.py
from helper_functions import get_relative_path


def calculate_priority_sum(line1, line2, line3):
    common = list(line1.intersection(line2, line3))[0]
    return (ord(common) - 38) if common.isupper() else (ord(common) - 96)


def main():
    priority_sum = 0

    with open(get_relative_path("input.txt")) as file:
        while line1 := set(file.readline().strip()):
            line2 = set(file.readline().strip())
            line3 = set(file.readline().strip())

            priority_sum += calculate_priority_sum(line1, line2, line3)

    print(priority_sum)


if __name__ == "__main__":
    main()

File: 2022/C++/Day-3/Part-2.cpp
#include <set>
#include <string>
#include <fstream>
#include <iostream>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::ifstream inputFile("input.txt");
    //-------------------------------------
    int priority_sum (0);

    std::string line1, line2, line3;
    while (std::getline(inputFile, line1) && !line1.empty()) {
           std::getline(inputFile, line2);
           std::getline(inputFile, line3);

        std::set<char> line1Set(line1.begin(), line1.end());
        std::set<char> line2Set(line2.begin(), line2.end());
        std::set<char> line3Set(line3.begin(), line3.end());

        for (char ch : line1Set) {
            if (line2Set.count(ch) && line3Set.count(ch)) {
                priority_sum += (isupper(ch) ? (ch - 38) : (ch - 96));
                break;
            }
        }
    }

    std::cout << priority_sum << "\n";
    //-------------------------------------
    return 0;
}